File: lo3common/src/main/scala/openapi/FicDetailsClient.scala
package kapibarabanka.lo3.common
package openapi

import models.domain.{Ao3FicInfo, Fic, FicDetails, Lo3Error, Note}

import kapibarabanka.lo3.common.models.api.FinishInfo
import scalaz.Scalaz.ToIdOps
import zio.http.*
import zio.http.Method.{GET, POST}
import zio.http.codec.*
import zio.http.endpoint.Endpoint

object FicDetailsClient extends MyClient:
  override protected val clientName = "fic-details"

  val patchDetails = (endpoint(Method.PATCH, "patch-details")
    .out[Unit]
    .outError[Lo3Error](Status.InternalServerError)
    |> withKey).in[FicDetails]

  val addNote = (endpoint(Method.PATCH, "add-note")
    .out[Unit]
    .outError[Lo3Error](Status.InternalServerError)
    |> withKey).in[Note]

  val startedToday = endpoint(Method.PATCH, "started-today")
    .out[Unit]
    .outError[Lo3Error](Status.InternalServerError)
    |> withKey

  val finishedToday = endpoint(Method.PATCH, "finished-today")
    .out[Unit]
    .outError[Lo3Error](Status.InternalServerError)
    |> withKey

  val abandonedToday = endpoint(Method.PATCH, "abandoned-today")
    .out[Unit]
    .outError[Lo3Error](Status.InternalServerError)
    |> withKey
  
  val finishFic = endpoint(POST, "finish-fic")
    .in[FinishInfo]
    .out[Unit]
    .outError[Lo3Error](Status.InternalServerError)

  override val allEndpoints =
    List(
      patchDetails,
      addNote,
      startedToday,
      finishedToday,
      abandonedToday,
      finishFic
    )

File: lo3web/src/lib/api/FicDetailsClient.ts
import type { FinishInfo } from '$lib/types/api-models';
import type { UserFicKey } from '$lib/types/domain-models';
import { getBaseUrl } from './api-utils';

const controller = 'fic-details';

export default class FicDetailsClient {
    public static startedToday(ficKey: UserFicKey): Promise<void> {
        return fetch(`${getBaseUrl()}/${controller}/started-today?` + new URLSearchParams({ ...ficKey }), {
            method: 'PATCH',
        }).then(() => {});
    }
    public static finishedToday(ficKey: UserFicKey): Promise<void> {
        return fetch(`${getBaseUrl()}/${controller}/finished-today?` + new URLSearchParams({ ...ficKey }), {
            method: 'PATCH',
        }).then(() => {});
    }
    public static abandonedToday(ficKey: UserFicKey): Promise<void> {
        return fetch(`${getBaseUrl()}/${controller}/abandoned-today?` + new URLSearchParams({ ...ficKey }), {
            method: 'PATCH',
        }).then(() => {});
    }
    public static finishFic(data: FinishInfo): Promise<void> {
        return fetch(`${getBaseUrl()}/${controller}/finish-fic`, {
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            method: 'POST',
            body: JSON.stringify(data),
        })
            .catch(e => console.log(e))
            .then(() => {});
    }
}
